src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -522,44 +524,66 @@ public static boolean checkBuiltInFunctionExists(FunctionIdentifier identifier)
             FunctionIdentifier identifier,
             List<RuntimeIterator> arguments,
+            ExecutionMode executionMode,
             IteratorMetadata metadata
     ) {
         BuiltinFunction builtinFunction = builtInFunctions.get(identifier);
 
         for (int i = 0; i < arguments.size(); i++) {
             if (!builtinFunction.getSignature().getParameterTypes().get(i).equals(mostGeneralSequenceType)) {
-                arguments.set(
-                    i,
-                    new TypePromotionIterator(
-                            arguments.get(i),
-                            builtinFunction.getSignature().getParameterTypes().get(i),
-                            "Invalid argument for function " + identifier.getName() + ". ",
-                            arguments.get(i).getMetadata()
-                    )
+                TypePromotionIterator typePromotionIterator = new TypePromotionIterator(
+                        arguments.get(i),
+                        builtinFunction.getSignature().getParameterTypes().get(i),
+                        "Invalid argument for function " + identifier.getName() + ". ",
+                        arguments.get(i).getHighestExecutionMode(),
+                        arguments.get(i).getMetadata()
                 );
+
+                arguments.set(i, typePromotionIterator);
             }
         }
 
         RuntimeIterator functionCallIterator;
         try {
-            Constructor<? extends RuntimeIterator> constructor = builtinFunction.getFunctionIterator()
+            Constructor<? extends RuntimeIterator> constructor = builtinFunction.getFunctionIteratorClass()
                 .getConstructor(
                     List.class,
                     IteratorMetadata.class
                 );
             functionCallIterator = constructor.newInstance(arguments, metadata);
         } catch (ReflectiveOperationException e) {
-            throw new UnknownFunctionCallException(identifier.getName(), arguments.size(), metadata);
+            // TODO: Some functions have constructors with 2 and some with 3 params. Handle this more elegantly
+            // TODO: Or alternatively, extend all constructors to have 3 params (executionMode)
+            try {
+                Constructor<? extends RuntimeIterator> constructor = builtinFunction.getFunctionIteratorClass()
+                        .getConstructor(
+                                List.class,
+                                ExecutionMode.class,
+                                IteratorMetadata.class
+                        );
+                functionCallIterator = constructor.newInstance(arguments, executionMode, metadata);
+            } catch (ReflectiveOperationException ex) {
+                throw new UnknownFunctionCallException(
+                        identifier.getName(),
+                        arguments.size(),
+                        metadata.getExpressionMetadata()
+                );
+            }
```

Comment:
I'd like to have your feedback on this one @ghislainfourny. To minimize the refactoring, I have initially opted for adding the execution mode as a parameter to the HybridRuntimeIterator and its descendants. LocalRuntimeIterators simply get this parameter fixed to LOCAL. This resulted in hybrid enabled built-in functions having constructors with 3 parameters while Local built-in functions have constructors with 2 parameters. In order to handle this and get the tests running. I have implemented the above fix which is quick and dirty.

One potential solution that would still be somewhat dirty but more performant. We could eliminate the exceptions by performing a certain check on the constructor to avoid performance loss due to exception handling.

Alternative feels more neat however will require a bit more refactoring. We can simply add ExecutionMode to all the constructors of all iterators and enforce that localRuntimeIterators can only be assigned ExecutionMode.LOCAL. A similar implementation can be found in DataFrameRuntimeIterator or RDDRuntimeIterator. 

What is your take on this? Do you see any preferable alternatives/suggestions?

Reply:
TODO - Offline discussion: expose executionMode parameters from all iterator constructors, remove the exception handling Above

Reply:
Done